### PR TITLE
"Easy Coding" GUI menu for activating loggers

### PR DESCRIPTION
**Pull request by ScegfOd**
here is the stuff to make a dialog that can activate logging from a gui menu (in advanced preferences)
At the moment it only prints to the command line that launched wesnoth, but I figured that would be fine for now.

### CODE REVIEW

File: data/gui/window/logging.cfg [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+#enddef
+#define _INFO
+"info"
+#enddef
+#define _WARN
+"warn"
+#enddef
+#define _ERR
+"err"
+#enddef
```
Comment (CelticMinstrel) [inline data/gui/window/logging.cfg]: I'm not sure what the point of these macros are. Just why do you need them? Anyway, if they really are necessary, the `#enddef` should not start on a new line.

File: data/gui/window/logging.cfg [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+		border = "all"
+		border_size = 5
+
+		[toggle_button]
+			id = {GROUP}
+			definition = "default"
+			linked_group = {GROUP}
+			tooltip = {TOOLTIP}
+			#wrap = "true"
```
Comment (CelticMinstrel) [inline data/gui/window/logging.cfg]: Please don't leave commented-out bits in your code unless there is a very good reason for them to be there.

File: src/gui/dialogs/logging.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    widget_id.push_back("err");
+    widget_id.push_back("warn");
+    widget_id.push_back("info");
+    widget_id.push_back("debug");
+
+    domain_list_end = "the end";
+
+    //empty string is the filter (in other words, this grabs the whole list of domains)
+    std::string temp_string = lg::list_logdomains("");
+    //std::cout<<temp_string;//use to print the full log domain list
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: This line isn't needed, right?

File: src/gui/dialogs/logging.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        set_logger(domain_list[counter]);
+    }
+}
+
+void tlogging::set_logger(const std::string log_domain)
+{
+	std::string active_value = groups_[log_domain].get_active_member_value();
+
+    if     (active_value == widget_id[1])//default value, level1: warning
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: What's with all this space between `if` and `(`?

File: src/gui/dialogs/logging.cpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            item["toggle_button"] = widget_id[iter];// id?
+            data.insert(std::make_pair(widget_id[iter], item));
+        }
+
+        /*to prevent all the individual toggles copying the label,
+         *(which will wreck everything)
+         *this widget MUST come last... I have no clue why
+         */
+		item["label"] = this_domain;
+		data.insert(std::make_pair("title_name", item));
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: `data["title_name"] = item` seems cleaner to me in this situation. I'd use insert only if I already have a pair from somewhere else.

**Comment by CelticMinstrel** [comment]
Okay, the major two things that need to be fixed before this can be merged are:

1. Fix the indentation in the C++ files - you're using spaces in some places and tabs in others. Tabs everywhere is preferred.
2. Add the new dialog to the unit tests.

The code style doesn't quite match with what Wesnoth is seemingly trying to use, but I don't personally consider this a blocker. Someone else might, however, so it's probably a good idea to at least fix the brace style to be more like this:

```C++
void f()
{
    if(a) {
    } else if(b) {
    } else {
    }
}
```

File: data/gui/window/logging.cfg (hunk 2 above)
Comment (ScegfOd) [inline data/gui/window/logging.cfg]: missed a spot, thanks
probably a leftover for when I was alternating between looking at other stuff and borrowing bits and testing what different things did xD

File: data/gui/window/logging.cfg (hunk 1 above)
Comment (ScegfOd) [inline data/gui/window/logging.cfg]: they're just to make it easier for the cpp and cfg to match, not really needed

**Comment by ScegfOd** [comment]
I think I'm done with the first thing, is adding the new dialog to the unit tests nontrivial?

**Comment by gfgtdf** [comment]
class memeber varaibles like tlogging::domain_list shoudl end with an underscore.

> I think I'm done with the first thing, is adding the new dialog to the unit tests nontrivial?

Adding ot to test ignore list is trivial. Note sure about how hard making a real test is.

**Comment by CelticMinstrel** [comment]
It's better to make a real test than adding it to the ignore list. Since this is not a very complicated dialog, you can probably copy and tweak an existing unit test? I don't actually know much about how the GUI2 unit tests work, though. Anyway, in my opinion, the ignore list should be restricted to dialogs that are too complicated to easily test... and even then it's better to try to make a test anyway.

File: data/gui/window/logging.cfg (hunk 1 above)
Comment (CelticMinstrel) [inline data/gui/window/logging.cfg]: I have no real problem with them being here, whether they are necessary or not, but I do think you should put the enddef on the same line, eg:

```wml
#define _WARN
"warn"#enddef
```

**Comment by CelticMinstrel** [comment]
I don't really see any significant problems now, though I might want to take a closer look at the WML later (it looks possibly a bit overcomplicated, but then, the GUI WML often does), and I haven't actually tested it yet.

**Comment by ScegfOd** [comment]
I tried a thing with what I'm pretty sure is the thing that tests gui2 stuff...
/me crosses fingers

**Comment by ScegfOd** [comment]
looks like what I did to test_gui2.cpp is good enough to add it to the unit tests... I think it only tests for if the window opens properly.

File: src/gui/dialogs/logging.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	set_restore(true); //why is this done manually?
+
+	tlistbox& logger_box = find_widget<tlistbox>(&window, "logger_listbox", false);
+
+	for(unsigned int counter = 0; domain_list_[counter] != domain_list_end_; counter++){
+		std::string this_domain = domain_list_[counter];
+		std::map<std::string, string_map> data;
+		string_map item;
+
+		for(unsigned int iter = 3; iter < 4; iter--){
```
Comment (aginor) [inline src/gui/dialogs/logging.cpp]: Please document/#define these magic numbers

File: src/gui/dialogs/logging.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		 *this widget MUST come last... I have no clue why
+		 */
+		item["label"] = this_domain;
+		data["label"] = item;
+
+		logger_box.add_row(data);
+		tgroup<std::string>& group = groups_[this_domain];
+
+		tgrid* this_grid = logger_box.get_row_grid(counter);
+		for(unsigned int iter = 0; iter < 4; iter++){
```
Comment (aginor) [inline src/gui/dialogs/logging.cpp]: And these... Why 4 in the first place? Is there a way you can make it dynamic?

**Comment by aginor** [comment]
It looks mostly good, please see the two comments I added. Could you please also add youself to the contributors and player_changelog? https://wiki.wesnoth.org/PatchSubmissionGuidelines explains how to do it.

I'd also appreciate it if you could add the new file to src/CMakeLists.txt.

We'd also appreciate it if you could rebase your PR to merge some of the 17 commits with identical titles.
Please look at https://wiki.wesnoth.org/DeveloperGuide#Commit_messages to see how we normally format our commit messages.

Thanks for contributing :)

**Comment by ScegfOd** [comment]
something more like that name? ^

**Comment by ScegfOd** [comment]
BTW, I haven't figured out how to rebase yet, should I just fork the updated master and copy the changes onto that ? I could fix the commit names too xD

File: data/core/about.cfg [hunk 8]
```diff
@@ -144,6 +144,10 @@
         comment = "coder, bug fixes"
     [/entry]
     [entry]
+        name = "Jonathan Combs (ScegfOd)"
```
Comment (aginor) [inline data/core/about.cfg]: Quoting from https://wiki.wesnoth.org/PatchSubmissionGuidelines
"New contributors (i.e., those without commit access) should add themselves to the Miscellaneous Contributors section of the credits file (data/core/about.cfg), making sure to follow alphabetical order with respect to the first character that will be displayed."

Please move your entry to the appropriate section.

File: src/gui/dialogs/logging.cpp (hunk 6 above)
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: Please don't use #define for numbers. Declare them as const int instead.

**Comment by aginor** [comment]
The easiest might be if you join the irc channel if you want help with rebasing. #wesnoth-dev on irc.freenode.net.

File: src/gui/dialogs/logging.cpp [hunk 9]
```diff
@@ -63,7 +63,8 @@ void tlogging::pre_show(twindow& window)
 		std::map<std::string, string_map> data;
 		string_map item;
 
-		for(unsigned int iter = (widget_id_.size()-1); iter < widget_id_.size(); iter--){
+		unsigned int id_size = widget_id_.size();
+		for(unsigned int iter = (id_size-1); iter < id_size; iter--){
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: This loop looks like it's asking for problems. It starts at 3, continues as long as `iter` is less than 4, and decrements the counter each time. Also, the name `iter` is misleading - it's a counter, not an iterator, so you should call it `i` or `j` or `n` or `m` or `counter` or something.

File: src/gui/dialogs/logging.cpp [hunk 10]
```diff
@@ -80,7 +81,7 @@ void tlogging::pre_show(twindow& window)
 		tgroup<std::string>& group = groups_[this_domain];
 
 		tgrid* this_grid = logger_box.get_row_grid(counter);
-		for(unsigned int iter = 0; iter < widget_id_.size(); iter++){
+		for(unsigned int iter = 0; iter < id_size; iter++){
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: Again, as mentioned above, this variable name is misleading - it's a counter, not an iterator.

**Comment by CelticMinstrel** [comment]
The rebasing is a minor issue, honestly - if you don't rebase it, I will when it's time to merge. However, if you want to learn how to do it, I'd be willing to help you out on IRC.

Since there are now conflicts which prevent Travis from running, I can't tell if the unit tests passed; did you run them to check?

**Comment by CelticMinstrel** [comment]
For some reason those two comments above claim to be on an outdated diff, but they do apply to the latest diff, so please address that.

**Comment by gfgtdf** [comment]
I wonder why you need the 'domain_list_end_' thing ?

File: src/gui/dialogs/logging.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	domain_list_.push_back(domain_list_end_);
+}
+
+void tlogging::pre_show(twindow& window)
+{
+	set_restore(true); //why is this done manually?
+
+	tlistbox& logger_box = find_widget<tlistbox>(&window, "logger_listbox", false);
+
+	for(unsigned int counter = 0; domain_list_[counter] != domain_list_end_; counter++){
```
Comment (CelticMinstrel) [inline src/gui/dialogs/logging.cpp]: I assume this is what @gfgtdf is referring to... look to where you use it, it does seem very weird. To begin with, assuming you actually want to iterate this way, this should probably be a const string in the source file and not exist at all in the header. Furthermore, this isn't really a good way to iterate through the logdomains.The `domain_list_` is a vector, so you can query its `.size()`, or if you don't need the counter for anything other than the line just below this, you can just use the range-for iteration syntax:

```c++
for(std::string this_domain : domain_list_)
```

That would replace this line and the next line, and eliminate the need for this weird `domain_list_end_` variable.

Comment (ScegfOd) [inline src/gui/dialogs/logging.cpp]: one of the loops needs to count down and the other needs access to the value of the counter for a grid getting method, but I like the look of that loop and I'll change the other two loops to match it

**Comment by ScegfOd** [comment]
I gave up trying to work around the errors git gave me so I opened a separate pull request from a properly made fork etc etc
sorry about that ^^;;

**Comment by cooljeanius** [comment]
> I gave up trying to work around the errors git gave me so I opened a separate pull request from a properly made fork etc etc
> sorry about that ^^;;

what was the number of this other separate pull request?

**Comment by ScegfOd** [comment]
Oh boy, what a blast from the past xD
640->646->650
I didn't know how to use git properly to rebase and force push back then xD
Think I should put a note at the end of PR 640?

**Comment by cooljeanius** [comment]
Adding number signs gets you cross-referencing:
#640 -> (this one) -> #650
(I came here because I was checking for other bugs about the `--logdomains` flag before filing #6018 btw)

**Comment by ScegfOd** [comment]
Sweet, thanks for the tip. I'll check out that #6018 Issue and see if I can help :grin:

**Pull request closed without merging**